### DISCUSSION TITLE
Block Coding Query - Attempting to get simple motor to mimic everything the main motor does

### DISCUSSION THREAD

**Discussion by EmperorKabir** [pybricks/support#1386] [Q&A] [answered]
I have a train set up with 2 city hubs with a train motor connected to each as shown. I am attempting to get one hub to simply mimic the actions of the first motor but I'm a coding n00b and struggling to find the correct blocks to make this happen. Note that I don't want to copy and paste a full set of motor instructions but instead do all my coding on one hub and just have the other copy whatever it happens to be doing.

I've seen the youtube tutorial but I'm unsure it works the same way for these train motors. N.B. the block coding below is very much half done, i.e. the get angle is irrelevant as is setting the motor to 50% on the receiving hub.

![image](https://github.com/pybricks/support/assets/114411808/a7b56229-80d2-4743-8c9d-98466b164303)

![image](https://github.com/pybricks/support/assets/114411808/0390d3b3-9d1a-4b2e-b9e4-61a84d73cb89)

If anyone could assist in what blocks to use and where, this would be very much appreciated!

**Comment by laurensvalk** [comment]
I was about to put a program together, but would it be easier to use one hub for both motors, and the other for the sensor?

The hub with the sensor could be the "main" hub, sending power commands to the other hub.

**Reply by EmperorKabir** [reply]
I was concerned about the power which is why I got a second hub at the same time as a second motor. Perhaps I didn't need to! In all honesty though I'd very much appreciate seeing both your suggestion and also the way I'm doing it as it will provide excellent reference material I can refer to to build up from.... If it's possible please!

**Reply by laurensvalk** [reply]
Working on an example as we speak :smile:

**Reply by laurensvalk** [reply]
In the following example I'm assuming both motors are on one hub. It could run a program like this.

It is just going to listen to the "power" value from the other hub and set both motors to that level:

![image](https://github.com/pybricks/support/assets/12326241/a6e3cf63-9fe7-4217-8c0c-8519a5ede77d)

This may be all that hub has to do. You can add all "intelligence" on the other end.

----------

That other hub, instead of controlling the motor directly, it's going to send the "power" value.

For example, here it's just going back and forth every 5 seconds, but you could do something interesting with the color sensor, like our train video.

![image](https://github.com/pybricks/support/assets/12326241/e2cb65b9-dc1a-431a-853a-f090cacfb993)

**Reply by laurensvalk** [reply]
I haven't tested this. Let me know how it goes!

**Reply by EmperorKabir** [reply]
Thank you! I'll test this later today, meanwhile I have one immediate question; in your first screenshot, you have a 50ms wait. Is this because it is impossible or impractical to make it continuously observed and a 50ms interval (which for practical purposes is as good as continuous) keeps it within hardware limitations?

**Reply by EmperorKabir** [reply]
Also is there no block coding option to broadcast the current motor power directly or I suppose it would have to be setting the main motor power to a variable rather than a number?

**Reply by laurensvalk** [reply]
The wait is technically optional. In this case, the motors will update 20 times per second, which is more than enough.

It's mainly a habit that can come in handy when you're doing a lot of other things at the same time.

**Reply by laurensvalk** [reply]
> Also is there no block coding option to broadcast the current motor power directly or I suppose it would have to be setting the main motor power to a variable rather than a number?

That's right. If you're asking about one motor on either hub, you could do it this way:

![image](https://github.com/pybricks/support/assets/12326241/7ca614a8-3fba-4ed9-9dd9-1c46d0f691b3)

There is always a slight communication delay, so in this way the motors may not start at exactly the same time, which may look less nice.

In the examples posted earlier, both motors always start at the same time. If there is a delay, they'll have the exact same delay so you'd be less likely to notice.

**Reply by EmperorKabir** [reply]
Hmmm I couldn't understand how the tasks work, in particular I couldn't work out how to bring in the yellow power button as in your example as it was not in the blocks that I could see. I tried a different way but am getting an error on the receiving motor:

![image](https://github.com/pybricks/support/assets/114411808/4e22ebcd-6222-4771-9137-bf6d70816ced)

**Reply by johnnyf64** [reply]
@ laurensvalk. What does the coding look like if I want to stop both motors upon color in your example.  Can anyone give an example. I'm not very familiar with coding yet. With one motor and one hub it's no problem.

**Reply by laurensvalk** [reply]
You can add an additional motor to the setup. Let's say you renamed the existing one `front motor` and the new one `rear motor`.

Likewise, duplicate each "set power" block, so that there is one for each motor.

**Reply by johnnyf64** [reply]
Thank you for the quick response. But how do I stop the two at the same time on a color scan, for example on the yellow color. Each motor needs its own color scan in a block to stop.

**Reply by laurensvalk** [reply]
If each motor needs its own color scan, I suppose that's a different setup than the one used here. Maybe you can open a new discussion, including a picture of your setup and a description of what you'd like to do?

**Reply by EmperorKabir** [reply]
It will also depend on the required scenario: 
1: one motor just replicates another - use my code here
2: one of either colour sensor seeing the colour(s) causes a new line of code
3: both colour sensors must see particular colour(s) at the same time for the next line of code to run
4: another scenario I've not described. 

You may need to specify and then if similar to my code, mention here or open a new thread

**Reply by johnnyf64** [reply]
Thanks both for the info. I'll try some things out. Otherwise I'll start a new thread. My setup is the same like https://www.youtube.com/watch?v=kwnSbnlhOkA&t=606s  I only want to use two motors on the train and two hubs. One for de colorsensor.

**Comment by EmperorKabir** [comment]
Aha, I think I cracked it!

![image](https://github.com/pybricks/support/assets/114411808/f678cc73-d961-4eb2-80bc-f7c915adfa4c)

**Reply by laurensvalk** [reply]
Well done standing by your original idea and making it work :smile:

**Reply by EmperorKabir** [reply]
Just as a side question, is there a tutorial on how tasks work? I've made a new version of this program which uses 2 colour sensors, one on each hub and the support hub simultaneously broadcasts back the colour it sees to the main hub (while receiving the power variable).

I'm doing this as at a hardware level, the colour sensor isn't the absolute best and the action condition is based on both colour sensors seeing the same thing to drastically reduce false positives.

Currently I'm using repeat loops within repeat loops to 'regularly rapidly' check the colours before moving on to the next command. I have a suspicion that using a task block may be a more efficient alternative but as a n00b, I'm not entirely sure how these really work, hence my question in the opening line. In my head, it's like running a simultaneous program alongside your main program but I'm not sure.

**Reply by laurensvalk** [reply, marked as answer]
As for color sensor accuracy, did you see the [tutorial on calibration](https://www.youtube.com/watch?v=kwnSbnlhOkA)?

We don't have tutorials for tasks, yet. Hopefully soon :)

**Reply by EmperorKabir** [reply]
Yes that was my starting point in fact but I still seem to get false positives. (I'm guessing you mean mapping in the floor/tracks as well due to the nature of how the program 'rounds' to the closest colour?)

I look forward to a tasks tutorial!

**Reply by laurensvalk** [reply]
That's right. I wonder if it might also see the train motor in your setup. Ideally it is placed a bit lower to the tracks, but maybe there isn't enough room?

**Reply by EmperorKabir** [reply]
Hmmm I see conflicting information on the internet. I've also seen that it shouldn't be too close and I've also tried to build a shroud around it with black Lego parts to minimise interference

**Reply by laurensvalk** [reply]
The way to tell is to print out the colors as in the video. The closer they get to "pure" blue, red, etc, the better. 

If they look like this, it's too far away:

![image](https://github.com/pybricks/support/assets/12326241/2e28d6b4-ea3e-4833-b107-3d429bbba95d)

This is from this [thread](https://github.com/orgs/pybricks/discussions/1335). But this works:

![image](https://github.com/pybricks/support/assets/12326241/1d4d4ff5-922d-4d8c-a2aa-99717ebe2032)
